Upgrade file storage abstraction layers

rijkvanzanten opened the pull request:
### Big Changes

- Replaces previous `@directus/drive` setup with a new (ESM-only) `@directus/storage` implementation
- Adds support for Cloudinary (through `@directus/storage-driver-cloudinary`)
- Lays the groundwork for enabling custom storage adapters
- Adds full test coverage for storage drivers
- Vastly reduces memory overhead on upload (by exclusively relying on streams, rather than reading the whole file into memory for metadata first[^1])
- Includes changes from https://github.com/directus/directus/pull/16424

### Left To-Do

- [x] Find exifr alternative / workaround to take Exif buffer directly rather than the whole file. See https://github.com/MikeKovarik/exifr/issues/101
- [x] Fix the tests in [api/src/storage/index.test.ts](https://github.com/directus/directus/pull/16825/files/05e4917f2161b9a3867831b9f479048847e32ccc#diff-2ef1a4c56f1a8e66b9fe7a763a54462a3fe9c88ef008a7e020d7aa0f7048a7ac) 

Opening as a draft as the above needs to be resolved before this is fully ready to merge, but I  could use some reviews and eyeballs on this in the meantime 🙂 

### Links

- https://github.com/directus/directus/discussions/15426
- https://github.com/directus/directus/discussions/14649
- https://github.com/directus/directus/discussions/10254
- https://github.com/directus/directus/discussions/3107
- https://github.com/directus/directus/discussions/14742
- https://github.com/directus/directus/discussions/14628
- https://github.com/directus/directus/discussions/3947
- https://github.com/directus/directus/discussions/14361
- https://github.com/directus/directus/discussions/5951
- Closes ENG-11
- Closes ENG-162
- Closes #16059
- Closes #16424

[^1]: This actually needs a hair more work 😁

rijkvanzanten committed:
Refactor storage logic to testable setup

rijkvanzanten committed:
Add tests to get storage config

rijkvanzanten committed:
Add tests for register drivers

rijkvanzanten committed:
Tests for index

rijkvanzanten committed:
Organize imports

rijkvanzanten committed:
Add missing getStorage

rijkvanzanten committed:
Setup boilerplate for cloudinary drive

rijkvanzanten committed:
Add build script

rijkvanzanten committed:
Add cloudinary configuration

rijkvanzanten committed:
WIP tweaks for R&D

rijkvanzanten committed:
Start storage abstraction v2

rijkvanzanten committed:
Make storage manager single file

rijkvanzanten committed:
Add test coverage

rijkvanzanten committed:
Organize imports

rijkvanzanten committed:
Setup local driver boilerplate

rijkvanzanten committed:
[WIP] Start on local driver 2.0

rijkvanzanten committed:
Add more methods

rijkvanzanten committed:
Lunchtime

rijkvanzanten committed:
Add put method

rijkvanzanten committed:
Add list method

rijkvanzanten committed:
[WIP] Try using storage in api

rijkvanzanten committed:
Use node16 module-resolution

rijkvanzanten committed:
Lets not mess with apply-query rn

rijkvanzanten committed:
I love CJS. Death to CJS.

rijkvanzanten committed:
Merge branch 'esm-interop' into feat/storage

rijkvanzanten committed:
Use dynamic imports

rijkvanzanten committed:
Merge branch 'main' into feat/storage

rijkvanzanten committed:
Make things work

rijkvanzanten committed:
Add path normalizer

rijkvanzanten committed:
Add azure storage driver

rijkvanzanten committed:
Update lock

rijkvanzanten committed:
Start on tests

rijkvanzanten committed:
Add getBuffer test

rijkvanzanten committed:
Add getStat tests

rijkvanzanten committed:
Add tests for exists

rijkvanzanten committed:
Add tests for move

rijkvanzanten committed:
Add tests for copy

rijkvanzanten committed:
Add tests for put

rijkvanzanten committed:
Add tests for delete

rijkvanzanten committed:
Add test coverage for list

rijkvanzanten committed:
Add removeLeading option to normalizePath

rijkvanzanten committed:
Use removeLeading option

rijkvanzanten committed:
Start on gcs

rijkvanzanten committed:
Add fullpath test

rijkvanzanten committed:
Add getStream

rijkvanzanten committed:
Add getBuffer

rijkvanzanten committed:
Add getStat

rijkvanzanten committed:
Add exists

rijkvanzanten committed:
Add move

rijkvanzanten committed:
Add copy

rijkvanzanten committed:
Add put

rijkvanzanten committed:
Add delete

rijkvanzanten committed:
Finish DriverGCS

rijkvanzanten committed:
Cleanup tests a bit

rijkvanzanten committed:
Start s3

rijkvanzanten committed:
Add getStream

rijkvanzanten committed:
Add getBuffer

rijkvanzanten committed:
Please Wryn

rijkvanzanten committed:
Add #exists

rijkvanzanten committed:
Use randomize data

rijkvanzanten committed:
No more hardcoded values 🙅‍♀️

rijkvanzanten committed:
Add tests for copy

rijkvanzanten committed:
Add tests for put

rijkvanzanten committed:
Add put/copy/delete

rijkvanzanten committed:
Add tests for delete

rijkvanzanten committed:
WIP list

rijkvanzanten committed:
Remove duplicate fullPath check

rijkvanzanten committed:
Finish AWS tests

rijkvanzanten committed:
Listen to wryn

rijkvanzanten committed:
Mission critical tweak

rijkvanzanten committed:
Add randomization, cleanup tests

rijkvanzanten committed:
Check src vs dest full path

rijkvanzanten committed:
Start on Cloudinary

rijkvanzanten committed:
Add parameter signature helper

rijkvanzanten committed:
Fix ESM building of shared

rijkvanzanten committed:
Fix ESM building of shared

rijkvanzanten committed:
Improve tests

rijkvanzanten committed:
Update pnpm

rijkvanzanten committed:
Remove old build commands

rijkvanzanten committed:
Merge branch 'main' into feat/storage

rijkvanzanten committed:
Merge branch 'shared-esm' into feat/storage

rijkvanzanten committed:
Generated d.ts files

rijkvanzanten committed:
Merge branch 'shared-esm' into feat/storage

rijkvanzanten committed:
Fix export naming

rijkvanzanten committed:
Merge branch 'shared-esm' into feat/storage

rijkvanzanten committed:
Move ESM only utils to new @directus/utils

rijkvanzanten committed:
Update lockfile

rijkvanzanten committed:
Fix type exports

rijkvanzanten committed:
Implement getStream

rijkvanzanten committed:
Cleanup tests

rijkvanzanten committed:
Simplify api

rijkvanzanten committed:
Simplify API

rijkvanzanten committed:
Simplify api

rijkvanzanten committed:
Simplify API

rijkvanzanten committed:
Add read/stat

rijkvanzanten committed:
Cleanup / add exists

rijkvanzanten committed:
Add move

rijkvanzanten committed:
Add write

rijkvanzanten committed:
Move uploadChunk to separate method

rijkvanzanten committed:
Add test for #uploadChunk

rijkvanzanten committed:
Add tests for write

rijkvanzanten committed:
Add copy

rijkvanzanten committed:
Add delete

rijkvanzanten committed:
Add list

rijkvanzanten committed:
Add list error handling

rijkvanzanten committed:
Remove old drive packages

rijkvanzanten committed:
Merge branch 'main' into feat/storage

rijkvanzanten committed:
Start updating API usage of storage

rijkvanzanten committed:
Use Readable instead of NodeJS.ReadableStream

rijkvanzanten committed:
Use readable instead of buffer

rijkvanzanten committed:
Restore shared to main

rijkvanzanten committed:
Update lockfile

rijkvanzanten committed:
Use a streaming pipeline for sharp metadata

rijkvanzanten commented:
@wrynegade just tagged you to take a glance at the test setups (based on the discussions we've had before) 🙂 

@freekrai / @ConnorSimply could y'all kick the tires for azure and gcs? 🙏🏻

freekrai commented:
> @freekrai / @ConnorSimply could y'all kick the tires for azure and gcs? 🙏🏻 

Can do / will do

paescuj commented on api/src/storage/index.test.ts:
```diff
@@ -0,0 +1,51 @@
+// import { StorageManager } from '@directus/drive';
```
This should probably go into the the To-Do list as well 🤓

rijkvanzanten replied:
Whoooops! Good one 🙂

paescuj commented on api/src/storage/register-drivers.ts:
```diff
@@ -0,0 +1,24 @@
+	const usedDrivers: string[] = [];
+
+	for (const [key, value] of Object.entries(env)) {
+		if ((key.startsWith('STORAGE') && key.endsWith('DRIVER')) === false) continue;
```
```suggestion
		if ((key.startsWith('STORAGE_') && key.endsWith('_DRIVER')) === false) continue;
```

paescuj commented on packages/storage-driver-azure/tsconfig.json:
```diff
@@ -0,0 +1,31 @@
+{
```
I very much welcome the fact that you're using the recommended config from https://github.com/tsconfig/bases/blob/main/bases/node18-strictest-esm.combined.json.
WDYT about going a step further by installing `@tsconfig/node18-strictest-esm` and extend the configs from this package?
```json
"extends": "@tsconfig/node18-strictest-esm/tsconfig.json"
```
That way things would be much clearer (as we can easily spot additional settings) and switching to a newer / other version would be a smoother story.

I know it's a bit off-topic but it would be a good time to do so now since this pull request affects 7 `tsconfig.json` files 😃
(Side note: This was also something I wanted to address in #9375 but wasn't aware of [tsconfig/bases](https://github.com/tsconfig/bases) at this time)

Also a bit finicky but you can omit the `exclude` property here since it defaults to "node_modules" and `outDir` anyway (see https://typescript-v2-121.ortam.vercel.app/docs/handbook/tsconfig-json.html#details).

rijkvanzanten replied:
Oh lovely! Didn't realize this config also existed as a pre-packaged thing 🙂 I used https://gist.github.com/rijkvanzanten/be696ff0d267b17dbe417bdb68cfc211 here now 🙂 

As for using another third party library, I think I might just package up the one we're using in our own `@directus/tsconfig`, just so we're gonna be able to easily make changes moving forwards (without having to add extend overrides)

paescuj replied:
> As for using another third party library, I think I might just package up the one we're using in our own `@directus/tsconfig`, just so we're gonna be able to easily make changes moving forwards (without having to add extend overrides)

Love it ❤️

rijkvanzanten replied:
https://github.com/directus/directus/pull/16825/commits/1795b8292990ca740b250726739364b009531b6b 🙌🏻

paescuj replied:
Awesome 🥇

paescuj commented:
Wow @rijkvanzanten, this is huge ❤️🚀🏅

>  - [ ] Find exifr alternative / workaround to take Exif buffer directly rather than the whole file. See Using exifr.parse on a buffer that contains just exif data MikeKovarik/exifr#101

I was about to come up with a PR addressing a few things related to the usage of `exifr`. (Coincidence? 😸)
Since I've noticed in this context that the library isn't being actively maintained at the moment, I was looking for some alternatives. The only one I could spot so far which looks promising is [ExifReader](https://github.com/mattiasw/ExifReader).
However, I've concluded that it would be better if we could stick to `exifr` because I still believe that it's a great library and moving to another library would most likely involve a lot of effort. Therefore I really hope we can come up with a workaround (see https://github.com/MikeKovarik/exifr/issues/101#issuecomment-1352453433) and I'd love to see it being somewhat maintained again.

azrikahar commented on packages/storage-driver-cloudinary/src/index.ts:
```diff
@@ -0,0 +1,387 @@
+		return new URLSearchParams(entries).toString();
+	}
+
+	/**
+	 * Generate the Cloudinary SHA1 signature for the given payload
+	 * @see https://cloudinary.com/documentation/signatures
+	 */
+	private getFullSignature(payload: Record<string, string>) {
+		const denylist = ['file', 'cloud_name', 'resource_type', 'api_key'];
+
+		const signaturePayload = Object.fromEntries(
+			Object.entries(payload).filter(([key]) => denylist.includes(key) === false)
+		);
+
+		const signaturePayloadString = this.toFormUrlEncoded(signaturePayload, { sort: true });
+
+		return createHash('sha1')
+			.update(signaturePayloadString + this.apiSecret)
+			.digest('hex');
+	}
+
+	/**
+	 * Creates inline URL signature for use with the image reading API
+	 * @see https://cloudinary.com/documentation/advanced_url_delivery_options#generating_delivery_url_signatures
+	 */
+	private getParameterSignature(filepath: string) {
+		return `s--${createHash('sha1')
+			.update(filepath + this.apiSecret)
+			.digest('base64url')
+			.substring(0, 8)}--`;
+	}
```
This is also more of a possible edge case concern, but do we want to make it possible/customizable to use `sha256` as well? It seems like it is technically possible for someone to limit their account to only allow SHA-256 (refer quote below), so I presume said users will face issues when the driver is only using sha1 hash. However I'm not sure how common that is among actual Cloudinary users.

> Note: By default, Cloudinary supports both SHA-1 and SHA-256 digests for validation, and you can use either. The SDK methods use the SHA-1 algorithm by default, but you can use the SHA-256 algorithm instead by setting the `signature_algorithm` SDK [configuration parameter](https://cloudinary.com/documentation/cloudinary_sdks#configuration_parameters) to `sha256`. If you want to limit your account to allow only the SHA-256 digest for all your validations, [submit a request](https://support.cloudinary.com/hc/en-us/requests/new).

Link to above quote in an info block: https://cloudinary.com/documentation/upload_images#manual_signature_generation

rijkvanzanten replied:
> However I'm not sure how common that is among actual Cloudinary users.

I think I feel the same way about this as the custom CDN url functionality. I don't know if this is common enough to warrant adding right now. This PR is already absolutely massive, so I'd rather launch this a little simpler, and circle back to that if the need ever comes up 🙂

rijkvanzanten replied:
You know what, nevermind! Turns out cloudinary supports both sha-1 and sha-256 by default, and you can optionally limit it to only sha-256 if you are enterprise. That means we should be able to use sha-256 at all times and make it work for everybody at all times 🙂 

> By default, Cloudinary supports both SHA-1 and SHA-256 digests for validation, and you can use either

https://cloudinary.com/documentation/signatures

azrikahar replied:
> That means we should be able to use sha-256 at all times and make it work for everybody at all times 🙂

That's actually way better and straightforward 🤯

azrikahar commented on packages/storage-driver-cloudinary/src/index.ts:
```diff
@@ -0,0 +1,387 @@
+		const fullPath = this.fullPath(filepath);
+		const resourceType = this.getResourceType(fullPath);
+
+		const timestamp = this.getTimestamp();
+
+		const uploadParameters = {
+			timestamp: timestamp,
```
```suggestion
		const uploadParameters = {
			timestamp: this.getTimestamp(),
```

https://github.com/directus/directus/blob/05e4917f2161b9a3867831b9f479048847e32ccc/packages/storage-driver-cloudinary/src/index.ts#L175

Super tiny suggestion to make it more similar to how timestamp parameter is directly set in the `move()` method.

rijkvanzanten replied:
You have a scary good eye for detail 😘 That's a leftover from an earlier state where the uploadchunk method was still within the same function context and it would reuse the function scope variable 🙂

azrikahar commented on packages/storage-driver-s3/package.json:
```diff
@@ -0,0 +1,44 @@
+		"access": "public"
+	},
+	"dependencies": {
+		"@aws-sdk/client-s3": "3.224.0",
```
Since this PR has also updated the s3 driver to use the latest v3 aws sdk, I think we can mark this to close #16059. However, we should also probably confirm this with an existing PR #16424 which aimed to do the same (and maybe close it in favor of this or carry over any implementations from there, if applicable).

rijkvanzanten replied:
Totally! I actually reused the changes in that PR. Can't believe I forgot to put it in the PR description 🙈 (Done now!)

azrikahar commented on packages/utils/src/normalize-path.ts:
```diff
@@ -0,0 +1,33 @@
+/**
+ * Replace windows style backslashes with unix forwards slashes
```
Understandably this is loosely based on the previous `normalize-path` package that is removed in this PR, but will we be using this outside of Node environment? 

Wondering can we also use `path.sep`/`path.posix.sep` in Node, similar to their usage in shared util function `pathToRelativeUrl`: 

https://github.com/directus/directus/blob/431f95210a5347a27e50661a2e35c789f161b381/packages/shared/src/utils/node/path-to-relative-url.ts#L4

but also understand that the current `\\`/`/` usages are technically what was already being used in the previous package.

rijkvanzanten replied:
> Wondering can we also use path.sep/path.posix.sep in Node, similar to their usage in shared node util function pathToRelativeUrl:

I don't think we can in this case! The goal of the function is to normalize windows style paths no matter what platform you're currently on. It's not trying to make the path "current environment" compatible, the goal is to turn any path into a unix path 🙂

azrikahar replied:
Makes sense! Initially this question was typed halfway before I realized it was based on the `normalize-path`, so just wanted to confirm the goal anyways 😄

licitdev committed:
Add basic e2e test for local upload and delete

licitdev committed:
Fix integer filesize in SQLite

br41nslug committed:
fixed environment mocking in unit tests

br41nslug committed:
Merge branch 'feat/storage' of https://github.com/directus/directus into feat/storage

br41nslug committed:
applied the same env mocking to other tests

rijkvanzanten committed:
Update api/src/storage/register-drivers.ts

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

paescuj commented:
Short update from my side concerning the metadata extraction: I've played around with the idea described in https://github.com/MikeKovarik/exifr/issues/101#issuecomment-1352453433. Although I've had some success (was able to inject the parser), I'm afraid this won't work since `exifr` expects the data in a different form. I think `sharpMetadata.exif` really only contains the EXIF metadata while `exifr` would expect all metadata...
 ```console
exifData {
  errors: [
     [Error: IFD0 offset points to outside of file.
     this.ifd0Offset: 19789, file.byteLength: 7248]
   ]
}

# ...

exifData { errors: [ [Error: Malformed EXIF data] ] }
```

I think we therefore have to take a different path.

@rijkvanzanten In https://github.com/MikeKovarik/exifr/issues/101 you've stated the following:

> [...] I do not have a Buffer with the full file contents available (nor am I able to get it) [...]

Wouldn't we, for example, be able to reuse the same stream which is passed to sharp? E.g. something similar to this (split pipe, without the writeable streams ofc.):

<table><tr><td><img width="500px" src=https://user-images.githubusercontent.com/5363448/207876622-8409026f-540c-4256-98f7-1e82bffb6987.png></td></tr><tr><td>Source: <a href="https://blog.insiderattack.net/a-visual-guide-to-nodejs-streams-9d2d594a9bf5">https://blog.insiderattack.net/a-visual-guide-to-nodejs-streams-9d2d594a9bf5</a></td></tr></table>

rijkvanzanten commented on packages/storage-driver-cloudinary/src/index.ts:
```diff
@@ -0,0 +1,387 @@
+		const resourceType = this.getResourceType(filepath);
+		const fullPath = this.fullPath(filepath);
+		const signature = this.getParameterSignature(fullPath);
+		const url = `https://res.cloudinary.com/${this.cloudName}/${resourceType}/upload/${signature}/${fullPath}`;
```
> However I do notice that there is a possibility that this can also be https://mycloud-res.cloudinary.com instead of https://res.cloudinary.com/mycloud, for private CDN distributions that is available for advanced plans or higher as mentioned in their docs: https://cloudinary.com/documentation/solution_overview#urls_and_endpoints
> 
> Not sure how edge-case-y that is, but maybe we can turn this into a private method or conditional statement to form the correct URL (maybe modifiable via cloudinary-driver-specific environment variable)

I think I'm gonna wait on this until the need arises. I don't know how common it is to have those custom slugs, and similarly to the sha-1 thing, it seems to be an enterprise enterprise only thing you have to reach out to support for in order to get "enabled" 🤔 Lets keep it simple for now, should be easy enough to add in later 👍🏻

paescuj commented:
> > Wouldn't we, for example, be able to reuse the same stream which is passed to sharp? E.g. something similar to this (split pipe, without the writeable streams ofc.):
> 
> Unfortunately not. Exifr only works with buffers, whereas the input is a stream. I do not want to turn that stream into a buffer, as that was the original memory problem 🙂 (Putting the stream data in a buffer would mean that we'd read the whole file into memory before passing it to exifr)

Of course! Sorry, I've missed that `exifr` isn't supporting streams as a input. But in this case it would be a possibility to extend `exifr` to support streams, right? I even think that would not be so difficult at all, since `exifr` reads the input in chunks anyway, see https://github.com/MikeKovarik/exifr/blob/ed816cbd4946bde0646fed31c5ff075d085c59bd/src/file-readers/ChunkedReader.mjs.
WDYT? Would you like me to take a closer look at this?

rijkvanzanten commented:
That should theoretically be possible, but I'm a tad worried that's way more work than we'd think. It is indeed funny how Exifr currently chunks the buffer and then parses every chunk, which is effectively the same as streaming 🤔 

I'm wondering if it's easier to pull out the part that actually parses the Exif data after extracting it from the file, and then use that to parse through the exif data buffer we already get from sharp 🤔

rijkvanzanten committed:
Use sha256 by default

paescuj commented:
> That should theoretically be possible, but I'm a tad worried that's way more work than we'd think. It is indeed funny how Exifr currently chunks the buffer and then parses every chunk, which is effectively the same as streaming 🤔
> 
> I'm wondering if it's easier to pull out the part that actually parses the Exif data after extracting it from the file, and then use that to parse through the exif data buffer we already get from sharp 🤔

Yeah, this would also be a possible approach. Although, I'm worried too, that it means quite some work as we would have to glue several components together (especially since we also need to detect which parsers / metadata segements are applicable for the image that is being processed). The first option therefore seems to be a bit more sustainable to me.
I think I might be able to come up with something in the next days - unless you have a strict deadline for this feature?

licitdev committed:
Add base read test for /assets

rijkvanzanten commented:
> Yeah, this would also be a possible approach. Although, I'm worried too, that it means quite some work as we would have to glue several components together (especially since we also need to detect which parsers / metadata segements are applicable for the image that is being processed). The first option therefore seems to be a bit more sustainable to me.
I think I might be able to come up with something in the next days (maybe until the end of the week) - unless you have a strict deadline for this feature?

I have hope! Exifr currently effectively already does that right? It scans through the file until it finds the exif block, and then starts parsing that. It continuous on, and then does the same for iptc / icc / etc. Sharp does that part for us at the moment, so I'm hoping we can somehow strip out the bit that scans through the file, in favor of passing the exif buffer straight from sharp to the correct logic in exifr that does the metadata conversion

paescuj commented:
> I have hope! Exifr currently effectively already does that right? It scans through the file until it finds the exif block, and then starts parsing that. It continuous on, and then does the same for iptc / icc / etc. Sharp does that part for us at the moment, so I'm hoping we can somehow strip out the bit that scans through the file, in favor of passing the exif buffer straight from sharp to the correct logic in exifr that does the metadata conversion

I've looked into it and I'm afraid that this would be very hard as `exifr` is simply not designed to be used in such a way.
Let me come up with a different approach until Sunday, okay?

rijkvanzanten committed:
Replace exifr with exif-reader

rijkvanzanten commented:
> I've looked into it and I'm afraid that this would be very hard as exifr is simply not designed to be used in such a way.
Let me come up with a different approach until Sunday, okay?

You're right! I just forked it and forked around a bit, but definitely not an easy one to modify in that way. That being said: https://github.com/directus/directus/pull/16825/commits/e418a42f7ff651a83ae1ddd0dc75f0676417f68a Problem solved! Turns out `sharp`'s tests use exif-reader, and after some more research it turns out the maintainer of `sharp` has push access to `exif-reader`, so it is up to date and works great with `sharp`'s exif buffer 😄

rijkvanzanten committed:
Fix tests for storage/index

rijkvanzanten committed:
Merge branch 'main' into feat/storage

freekrai reviewed (approved):
GCS testing went smooth, uploaded several files and worked nicely.

Also uploaded a 497MB video file and that seemed to go quite a bit faster as well.

rijkvanzanten committed:
Install faking lib for tests

rijkvanzanten committed:
Add test for register-drivers

rijkvanzanten committed:
Add tests for register-locations

rijkvanzanten committed:
Finish tests

rijkvanzanten committed:
Organize imports

rijkvanzanten commented on packages/storage-driver-cloudinary/src/index.ts:
```diff
@@ -0,0 +1,387 @@
+			chunks.push(chunk);
+			currentChunkSize += chunk.length;
+
+			if (currentChunkSize >= 5.5e6) {
```
I'll add a comment 🙂

rijkvanzanten commented on packages/storage-driver-cloudinary/src/index.ts:
```diff
@@ -0,0 +1,387 @@
+		const fullPath = this.fullPath(filepath);
+		const resourceType = this.getResourceType(fullPath);
+		const publicId = this.getPublicId(fullPath);
+		const url = `https://api.cloudinary.com/v1_1/${this.cloudName}/resources/${resourceType}/upload/${publicId}`;
```
I think i'll leave it as-is, seeing that methods might change over time independent of endpoint 🙂

rijkvanzanten committed:
Resolve Azris comments

rijkvanzanten committed:
Fix outdated tests

Certainly not the cleanest, but I need to do a bigger pass on all these tests to get them up to date with the latest team requirements. Gonna do that in a separate PR..

rijkvanzanten committed:
Test for sha256

rijkvanzanten committed:
Attempt 1 at fixing toString errorr

I'm not seeing this issue locally, so we'll spam a couple commits here to get things going

rijkvanzanten committed:
Use node 18 in tests?!

rijkvanzanten commented:
@licitdev Could you help me figure out what's preventing the blackbox tests from running? 🙏🏻

connorwinston reviewed (approved):
Tested using Azure Blob Container storage and looked good to me. Uploaded, edited, and deleted files! 👍

licitdev committed:
Fix localhost resolution with 127.0.0.1

licitdev reviewed (commented):
Node.js favours IPv6 addresses in v17+, which resulted in `localhost` being resolved as `::1`. Ref: https://github.com/nodejs/node/issues/40537

licitdev committed:
Merge branch 'main' into feat/storage

licitdev committed:
Mock getEnv()

rijkvanzanten commented:
@licitdev you're a legend! Thanks for figuring that out 💪🏻

rijkvanzanten committed:
Merge branch 'main' into feat/storage

rijkvanzanten committed:
Use @directus/tsconfig instead of duplicated conf

rijkvanzanten committed:
Does this fix it?

rijkvanzanten committed:
OK fun detour

rijkvanzanten committed:
Recreate lockfile

rijkvanzanten committed:
Update config files

rijkvanzanten committed:
Use multipart uploads in S3

rijkvanzanten committed:
Cleanup imports

paescuj committed:
File Storage 2.0: Make metadata extraction backward-compatible (#16868)

rijkvanzanten commented:
@licitdev Could you help me figure out what's preventing the blackbox tests from running? 🙏🏻 

hihi i borked it again. Not quite sure how the change affects esbuild as well, but maybe you spot some weird dependency thing......

paescuj commented:
> @licitdev Could you help me figure out what's preventing the blackbox tests from running? 🙏🏻

@licitdev FYI: I came up with a fix in #16871

paescuj committed:
Reinstall packages using pnpm instead of manually removing them (#16871)

br41nslug committed:
Added extra environment setting for sharp processing of invalid images (#16811)

* Added extra environment setting for sharp processing of invalid images

* renamed environment var to `ASSETS_INVALID_IMAGE_SENSITIVITY_LEVEL`

paescuj commented on packages/storage-driver-azure/package.json:
```diff
@@ -0,0 +1,46 @@
+	"author": "Rijk van Zanten <rijkvanzanten@me.com>",
+	"exports": {
+		".": "./dist/index.js",
+		"./node": "./dist/node/index.js",
```
Probably a leftover?
```suggestion
```

paescuj replied:
Probably a leftover?
```suggestion
```

paescuj commented on packages/storage-driver-azure/tsconfig.json:
```diff
@@ -0,0 +1,8 @@
+		"outDir": "dist"
+	},
+	"include": ["src"],
+	"exclude": ["dist", "node_modules"]
```
Not required since it is the default
```suggestion
```

paescuj commented on packages/storage-driver-cloudinary/tsconfig.json:
```diff
@@ -0,0 +1,8 @@
+		"outDir": "dist"
+	},
+	"include": ["src"],
+	"exclude": ["dist", "node_modules"]
```
```suggestion
```

paescuj commented on packages/storage-driver-gcs/readme.md:
```diff
@@ -0,0 +1,3 @@
+# `@directus/storage-driver-azure`
+
+Azure file storage driver for `@directus/storage`
```
```suggestion
# `@directus/storage-driver-gcs`

GCS file storage driver for `@directus/storage`
```

paescuj commented on packages/storage-driver-gcs/tsconfig.json:
```diff
@@ -0,0 +1,8 @@
+		"outDir": "dist"
+	},
+	"include": ["src"],
+	"exclude": ["dist", "node_modules"]
```
```suggestion
```

paescuj commented on packages/storage-driver-local/tsconfig.json:
```diff
@@ -0,0 +1,8 @@
+		"outDir": "dist"
+	},
+	"include": ["src"],
+	"exclude": ["dist", "node_modules"]
```
```suggestion
```

paescuj commented on packages/storage-driver-s3/tsconfig.json:
```diff
@@ -0,0 +1,8 @@
+		"outDir": "dist"
+	},
+	"include": ["src"],
+	"exclude": ["dist", "node_modules"]
```
```suggestion
```

paescuj commented on packages/storage/tsconfig.json:
```diff
@@ -0,0 +1,8 @@
+		"outDir": "dist"
+	},
+	"include": ["src"],
+	"exclude": ["dist", "node_modules"]
```
```suggestion
```

paescuj commented on packages/utils/package.json:
```diff
@@ -0,0 +1,51 @@
+	"funding": "https://github.com/directus/directus?sponsor=1",
+	"license": "GPL-3.0",
+	"author": "Rijk van Zanten <rijkvanzanten@me.com>",
+	"exports": {
+		".": {
+			"import": "./dist/index.js",
+			"types": "./dist/index.d.ts"
+		},
+		"./node": {
+			"import": "./dist/node/index.js",
+			"types": "./dist/node/index.d.ts"
+		},
+		"./package.json": "./package.json"
+	},
```
According to the [TS docs](https://www.typescriptlang.org/docs/handbook/esm-node.html#packagejson-exports-imports-and-self-referencing) the "types" property must occur first:
```jsonc
    "exports": {
        ".": {
            // Entry-point for TypeScript resolution - must occur first!
            "types": "./types/index.d.ts",
            // Entry-point for `import "my-package"` in ESM
            "import": "./esm/index.js",
            // Entry-point for `require("my-package") in CJS
            "require": "./commonjs/index.cjs",
        },
    },
```

Not sure if this only applies when having both "import" and "require", but better be safe:

```suggestion
	"exports": {
		".": {
			"types": "./dist/index.d.ts",
			"import": "./dist/index.js"
		},
		"./node": {
			"types": "./dist/node/index.d.ts",
			"import": "./dist/node/index.js"
		},
		"./package.json": "./package.json"
	},
```

rijkvanzanten replied:
TypeScript never seizes to amaze me 🥲 I guess it makes sense though! It'll go top to bottom figuring out which thing to use, and short-circuits on the first hit. Good catch!

paescuj replied:
> TypeScript never seizes to amaze me 🥲

You have taken the words out of my mouth 🤣😢

> I guess it makes sense though! It'll go top to bottom figuring out which thing to use, and short-circuits on the first hit. Good catch!

Yes, I think so too.

paescuj commented on packages/utils/readme.md:
```diff
@@ -0,0 +1,3 @@
+# `@directus/storage-driver-azure`
+
+Azure file storage driver for `@directus/storage`
```
```suggestion
# `@directus/utils`

Utilities shared between the Directus packages
```

paescuj commented on packages/utils/tsconfig.json:
```diff
@@ -0,0 +1,8 @@
+		"outDir": "dist"
+	},
+	"include": ["src"],
+	"exclude": ["dist", "node_modules"]
```
```suggestion
```

paescuj reviewed (approved):
Just a few small things... Apart from that, we are ready to go from my point of view 🚀

rijkvanzanten committed:
Remove unused excludes from tsconfig

rijkvanzanten committed:
Remove copy/paste leftover

rijkvanzanten committed:
Update packages/utils/readme.md

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

rijkvanzanten committed:
Update packages/utils/package.json

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

rijkvanzanten committed:
Merge branch 'main' into feat/storage

abdonrd commented:
After update to v9.22.0, we see this error:

```
[21:46:50.566] ERROR: storage:s3:responseTime in ERROR state
    err: {
      "type": "Error",
      "message": "Region is missing",
      "stack":
          Error: Region is missing
              at default (/directus/node_modules/@aws-sdk/client-s3/node_modules/@aws-sdk/config-resolver/dist-cjs/regionConfig/config.js:10:15)
              at /directus/node_modules/@aws-sdk/client-s3/node_modules/@aws-sdk/node-config-provider/dist-cjs/fromStatic.js:6:83
              at /directus/node_modules/@aws-sdk/client-s3/node_modules/@aws-sdk/property-provider/dist-cjs/chain.js:11:28
              at async coalesceProvider (/directus/node_modules/@aws-sdk/client-s3/node_modules/@aws-sdk/property-provider/dist-cjs/memoize.js:14:24)
              at async /directus/node_modules/@aws-sdk/client-s3/node_modules/@aws-sdk/property-provider/dist-cjs/memoize.js:26:28
              at async useFipsEndpoint (/directus/node_modules/@aws-sdk/client-s3/node_modules/@aws-sdk/config-resolver/dist-cjs/regionConfig/resolveRegionConfig.js:21:74)
              at async resolveParams (/directus/node_modules/@aws-sdk/client-s3/node_modules/@aws-sdk/middleware-endpoint/dist-cjs/adaptors/getEndpointFromInstructions.js:29:40)
              at async getEndpointFromInstructions (/directus/node_modules/@aws-sdk/client-s3/node_modules/@aws-sdk/middleware-endpoint/dist-cjs/adaptors/getEndpointFromInstructions.js:7:28)
              at async /directus/node_modules/@aws-sdk/client-s3/node_modules/@aws-sdk/middleware-endpoint/dist-cjs/endpointMiddleware.js:8:26
              at async /directus/node_modules/@aws-sdk/client-s3/node_modules/@aws-sdk/middleware-logger/dist-cjs/loggerMiddleware.js:5:22
    }
```

rijkvanzanten commented:
@abdonrd Yeah seems to be some discrepancies in when region is required vs when it's not between vendors. https://github.com/directus/directus/pull/16877 will go out to resolve that shortly 👍🏻

abdonrd commented:
Thanks for the quick fix, @rijkvanzanten!

paescuj commented:
> Ouch! It seems that we still have the same error with `v9.22.1`. We are going to downgrade and try tomorrow.

Does this help?: https://github.com/directus/directus/issues/16876#issuecomment-1362269999

StephenGunn commented:
> Ouch! It seems that we still have the same error with `v9.22.1`. We are going to downgrade and try tomorrow.

Try adding https:// to your endpoint

abdonrd commented:
We finally add the region, and it's working now.

abdonrd commented:
On another project that we have **where we are not using File Storage**, we have this error:

```
INFO [00:05:14.510] INFO: Loaded extensions: ***, ***
WARN [00:05:14.525] WARN: Upload directory (/directus/uploads) is not read/writeable!
WARN [00:05:14.604] WARN: PostGIS isn't installed. Geometry type support will be limited.
INFO [00:05:14.898] INFO: Server started at http://0.0.0.0:8055
[00:05:15] GET /server/health 200 104ms
node:events:491
      throw er; // Unhandled 'error' event
      ^
Error Error: ENOENT: no such file or directory, open '/directus/uploads/health-rRsld'
Emitted 'error' event on ReadStream instance at:
    at emitErrorNT (node:internal/streams/destroy:157:8)
    at emitErrorCloseNT (node:internal/streams/destroy:122:3)
    at processTicksAndRejections (node:internal/process/task_queues:83:21) {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '/directus/uploads/health-rRsld'
}
```

rijkvanzanten commented:
@abdonrd mind making a new issue? Merged PR comments are really hard to track / triage :)
